query/src/main/java/org/infinispan/query/clustered/DistributedIterator.java

```diff
@@ -99,6 +105,78 @@ public final T next() {
          throw new NoSuchElementException();
       }
 
+      // hasNext populate the values if returns true
+      assert !values.isEmpty();
+      assert valueIndex < values.size();
+
+      return values.get(valueIndex++);
+   }
+
+   /**
+    * Extension point for subclasses.
+    */
+   protected T decorate(Object key, Object value, float score) {
+      return (T) value;
+   }
+
+   @Override
+   public final boolean hasNext() {
+      if (valueIndex == values.size()) {
+         fetchBatch();
+      }
+      return valueIndex < values.size();
+   }
+
+   private void fetchBatch() {
+      // keep the order
+      List<KeyAndScore> keys = new ArrayList<>(batchSize);
+      values.clear();
+      valueIndex = 0;
+      for (int i = 0; i < batchSize; ++i) {
+         if (!hasMoreKeys()) {
+            break;
+         }
+         KeyAndScore key = nextKey();
+         if (key != null) {
+            keys.add(key);
+         }
+      }
+
+      if (keys.isEmpty()) {
+         return;
+      }
+
+      if (!queryStatistics.isEnabled()) {
+         getAllAndStore(keys);
+         return;
+      }
+
+      TimeService timeService = SecurityActions.getCacheComponentRegistry(cache).getTimeService();
+      long start = timeService.time();
+      getAllAndStore(keys);
+      queryStatistics.entityLoaded(timeService.timeDuration(start, TimeUnit.NANOSECONDS));
```

Comment:
at the moment the `entityLoaded` method is supposed to measure 1 load, so we should count the fact that with this change (luckily) we're going to fetch several entities with one operation

Reply:
I was planning to change it but I checked the other place in the code and it not counting the number of keys, just the requests:

https://github.com/infinispan/infinispan/blob/b78b318de38ae32ca0e97bfd7f397d50d6fd3e2e/query/src/main/java/org/infinispan/query/impl/EntityLoader.java#L40-L45

Reply:
Hi Pedro. You're right. Please ignore this comment.